File: heraldus.py
import logging
import sys
import imaplib
import email
import email.message
from email.header import decode_header
from datetime import datetime, date
from telegram import Update
from telegram.ext import Updater, CommandHandler, MessageHandler, Filters, CallbackContext


logging.basicConfig(level=logging.INFO,
                    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')


class UserMail:
    def __init__(self, user, password):
        self.user = user
        self.password = password


mail_boxes = {}


def get_mail_server(user_mail):
    try:
        mail = imaplib.IMAP4('localhost', 143)
        mail.login(user_mail.user, user_mail.password)
        mail.select('INBOX')
        return mail
    except imaplib.IMAP4.error:
        return None


def set_check_email(update: Update, context: CallbackContext):
    if not update.message:
        return

    user_mail = mail_boxes.get(update.message.chat.id)
    if len(context.args) != 2 or user_mail or not get_mail_server(UserMail(*context.args)):
        return

    logging.info(context.args)
    j = updater.job_queue
    j.run_repeating(check_email, interval=300, context={'chat': update.message.chat.id, 'args': context.args})
    mail_boxes[update.message.chat.id] = UserMail(*context.args)
    update.message.reply_text('done!')


def mark_msgs_as_read(update, context):
    user_mail = mail_boxes.get(update.message.chat.id)
    if user_mail:
        mail = get_mail_server(user_mail.user)
        type, data = mail.search(None, 'UNSEEN')
        for msg in data:
            mail.store(msg.replace(' ', ','), '+FLAGS', '\Seen')
        update.message.reply_text('clear!')


def echo(update: Update, context: CallbackContext) -> None:
    if update.message.chat.id not in mail_boxes:
        update.message.reply_text("You aren't subscribed to mail")
    else:
        update.message.reply_text('Great! You are subscribed to mail')


def get_header(header):
    decode_subject = decode_header(header)
    if decode_subject[0][1]:
        return decode_subject[0][0].decode(decode_subject[0][1])
    return decode_subject[0][0]


def check_email(context: CallbackContext):
    job = context.job
    user_mail = mail_boxes.get(job.context['chat'])
    if user_mail:
        mail = get_mail_server(user_mail)
        status, response = mail.search(None, 'UNSEEN')
        unread_msg_nums = response[0].split()
        response = ''
        for msg_num in unread_msg_nums:
            status, msg_data = mail.fetch(msg_num.decode('utf-8'), '(RFC822)')
            
            email_msg = email.message_from_bytes(msg_data[0][1], _class = email.message.EmailMessage)
            response += '{}. {} - {}\n'.format(msg_num.decode('utf-8'),
                    get_header(email_msg['From']),
                    get_header(email_msg['Subject']))

        if len(response):
            context.bot.send_message(job.context['chat'], text=response)


if __name__ == '__main__':
    updater = Updater(token='')
    dispatcher = updater.dispatcher

    dispatcher.add_handler(CommandHandler('set_check_email', set_check_email))
    dispatcher.add_handler(MessageHandler(Filters.text & ~Filters.command, echo))
    
    updater.start_polling()
    updater.idle()

File: heraldus.go
package main

import (
    "fmt"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/emersion/go-imap"
	imapclient "github.com/emersion/go-imap/client"
	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
)

type EmailUser struct {
    Email string
    Password string
}

func login_user(email string, password string) (*imapclient.Client, error) {
    log.Printf("login user: %s %s ", email, password)

    c, err := imapclient.Dial("localhost:143")
	if err != nil {
		log.Fatal(err)
        return nil, err
	}

    err = c.Login(email, password)
    if err != nil {
        log.Panic(err)
        return nil, err
    }

    return c, nil
}

func fetch(user *EmailUser) []string {
    c, err := login_user(user.Email, user.Password)
	if err != nil {
        log.Println(err)
        return []string{}
	}

    log.Println("Connected")
    defer c.Logout()

    _, err = c.Select("INBOX", false)
    if err != nil {
        log.Println(err)
        return []string{}
    }
    log.Println("Selected INBOX")

    criteria := imap.NewSearchCriteria()
    criteria.WithoutFlags = []string{"\\Seen"}
    uids, err := c.Search(criteria)
    if err != nil {
        log.Println(err)
        return []string{}
    }

    seqset := new(imap.SeqSet)
    seqset.AddNum(uids...)
    items := []imap.FetchItem{imap.FetchEnvelope}
    messages := make(chan *imap.Message)
    done := make(chan error, 1)
    go func() {
        done <- c.Fetch(seqset, items, messages)
    }()

    emails := make([]string, 0, len(uids))
    for msg := range messages {
        log.Println(msg.Envelope.Subject)
        emails = append(emails, fmt.Sprintf("%s - %s", msg.Envelope.From[0].PersonalName, msg.Envelope.Subject))
    }

    // mark seen messages
    seenFlagItem := imap.FormatFlagsOp(imap.AddFlags, true)
    flags := []interface{}{imap.SeenFlag}
    err = c.Store(seqset, seenFlagItem, flags, nil)
    if err != nil {
        log.Println(err)
    }

    return emails
}

func main() {
    users := map[int64]EmailUser{}
    mu := sync.Mutex{}

    bot, err := tgbotapi.NewBotAPI("")
    if err != nil {
        log.Panic(err)
    }

    bot.Debug = true
    
    log.Printf("Authorized on account %s", bot.Self.UserName)

    u := tgbotapi.NewUpdate(0)
    u.Timeout = 60

    updates := bot.GetUpdatesChan(u)

    go func(users* map[int64]EmailUser) {
        for {
            mu.Lock()

            log.Println("Go checking emails")
            for chatId, email := range *users {
                letters := fetch(&email)
                if len(letters) == 0 {
                    continue
                }
                reply := strings.Join(letters, "\n")
                msg := tgbotapi.NewMessage(chatId, reply)
                bot.Send(msg)
            }

            mu.Unlock()
            time.Sleep(1 * time.Minute)
        }
    }(&users)

    for update := range updates {
        if update.Message != nil {
            chatId := update.Message.Chat.ID
            if update.Message.IsCommand() {
                switch update.Message.Command() {
                case "set_check_email":
                    args := strings.Split(update.Message.CommandArguments(), " ")
                    if len(args) != 2 {
                        continue
                    }
                    
                    _, ok := users[chatId]
                    if ok {
                        continue
                    }

                    c, err := login_user(args[0], args[1])
                    c.Logout()

                    if err == nil {
                        mu.Lock()
                        users[chatId] = EmailUser{args[0], args[1]}
                        mu.Unlock()
                        msg := tgbotapi.NewMessage(chatId, "Done!")
                        msg.ReplyToMessageID = update.Message.MessageID
                        bot.Send(msg)
                    }
                }
            } else if len(update.Message.Text) > 0 {
                var reply string
                mu.Lock()
                _, ok := users[chatId]
                mu.Unlock()
                if ok {
                    reply = "You are already subscribed to mail!"
                } else {
                    reply = "You aren't subscribed!"
                }

                msg := tgbotapi.NewMessage(update.Message.Chat.ID, reply)
                bot.Send(msg)
            }
        }
    }
}
